### DISCUSSION TITLE
What channel / which cc are used?

### DISCUSSION THREAD

**Discussion by marcel-licence** [marcel-licence/esp32_basic_synth#3] [Q&A]
Playing notes are possible on all channels. There is no check implemented
-> ref. void HandleShortMsg(uint8_t *data)

Changing parameters are associated to the contol change message (0xb0):
-> ref. void Midi_ControlChange(uint8_t channel, uint8_t data1, uint8_t data2)

Because I had a weird setup of my midi controller the mapping is a bit strange.
You can adopt the implementation for your own needs by calling:
- Synth_SetRotary
- Synth_SetSlider
when using specific messages. Data1 refers to the cc you want to use. Data2 is the value transmitted by your controller.

**Comment by Opekusha** [comment]
I figured it out. Many thanks. Success and new great ideas.

**Comment by coatsiecoates** [comment]
I'm having difficulty getting full use of MIDI CCs. 

If I specify 0x0, and then the proposed CC number in z_config:

```
    { 0x0, 0x1, "R1", NULL, Synth_SetParam, SYNTH_PARAM_DETUNE_1},//1
    { 0x0, 0x2, "R2", NULL, Synth_SetParam, SYNTH_PARAM_UNISON_2},//2
    { 0x0, 0x3, "R3", NULL, Delay_SetLength, 2},//3
    { 0x0, 0x4, "R4", NULL, Delay_SetOutputLevel, 3},//4
    //{ 0x0, 0x5, "R5", NULL, Delay_SetFeedback, 4},//5
    { 0x0, 0x6, "R6", NULL, Synth_SetParam, SYNTH_PARAM_MAIN_FILT_CUTOFF},//6
    { 0x0, 0x7, "R7", NULL, Synth_SetParam, SYNTH_PARAM_MAIN_FILT_RESO},//7
    //{ 0x7, 0x10, "R8", NULL, Synth_SetParam, SYNTH_PARAM_VOICE_FILT_RESO},
    { 0x0, 0x8, "R8", NULL, Synth_SetParam, SYNTH_PARAM_VOICE_PORT_TIME},//8
    { 0x0, 0x9, "R9", NULL, Reverb_SetLevel, SYNTH_PARAM_VOICE_NOISE_LEVEL},//9

```
I can use CCs to control them.

But if I do the same for the ADSR controls, nothing happens:

```
    { 0x0, 0xA, "S1", NULL, Synth_SetParam, SYNTH_PARAM_VEL_ENV_ATTACK},//10
    { 0x0, 0xB, "S2", NULL, Synth_SetParam, SYNTH_PARAM_VEL_ENV_DECAY},//11
    { 0x0, 0xC, "S3", NULL, Synth_SetParam, SYNTH_PARAM_VEL_ENV_SUSTAIN},//12
    { 0x0, 0xD, "S4", NULL, Synth_SetParam, SYNTH_PARAM_VEL_ENV_RELEASE},//13
    { 0x0, 0x5, "S5", NULL, Synth_SetParam, SYNTH_PARAM_FIL_ENV_ATTACK},//14
    { 0x0, 0xF, "S6", NULL, Synth_SetParam, SYNTH_PARAM_FIL_ENV_DECAY},//15
    { 0x0, 0x10, "S7", NULL, Synth_SetParam, SYNTH_PARAM_FIL_ENV_SUSTAIN},//16
    { 0x0, 0x11, "S8", NULL, Synth_SetParam, SYNTH_PARAM_FIL_ENV_RELEASE},//17
```
Is it possible to access all parameters on a given channel simply using incoming CCs?

Are there docs that explain Synth_SetRotary and Synth_SetSlider? Why does rotary or slider matter? Can I just bypass that and send only the CC number and value somewhere?

**Reply by marcel-licence** [reply]
Hey,
it doesn't matter if it is rotary or slider. I just used these functions to get a better overview when working with a specific controller. The map would be enough using direct callback functions. In some cases you would like to avoid writing a function for each features. 

At the moment there is a callback for raw messages ref. https://github.com/marcel-licence/ML_SynthTools/blob/main/src/midi_interface.h
line 177.

You could check the bytes to detect the CC messages and process them. 

Hope this helps.

**Comment by coatsiecoates** [comment]
oh, that's perfect! I look forward to playing around with this.

**Comment by coatsiecoates** [comment]
...and thank you... Forgive my manners.
